rewrite-java/src/main/java/org/openrewrite/java/ReplaceAnnotation.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.*;
+import org.openrewrite.java.search.UsesType;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.TypeUtils;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class ReplaceAnnotation extends Recipe {
+    @Option(displayName = "Annotation pattern to replace",
+            description = "An annotation pattern, expressed as a method pattern to replace.",
+            example = "@org.jetbrains.annotations.NotNull(\"Test\")")
+    String annotationPatternToReplace;
+    @Option(displayName = "Annotation template to insert",
+            description = "An annotation template to add instead of original one, will be parsed with `JavaTemplate`.",
+            example = "@NonNull")
```

Comment:
I think it would make sense with an example that includes a parameter, to make it clear that this is supported. Additionally, since a parameter could require a reference to a Java class, I think it would make sense for the JavaTemplate here to be required to be fully context-free and thus not rely on any imports being registered with the JavaTemplate. As a result I imagine the `annotationFQN` parameter can be dropped (didn't look at the code). Then we can apply the `ShortenFullyQualifiedNames` visitor to get imports added.

Reply:
I tried executing `ShortenFullyQualifiedNames` with `doAfterVisit(new ShortenFullyQualifiedTypeReferences().getVisitor());` earlier, but this didn't make the expected changes. I'll try again and post a gist if not working.

Reply:
IIRC the class declares a static utility method. We should find some uses of it in rewrite-static-analysis.

Reply:
With `ShortenFullyQualifiedTypeReferences#modifyOnly` the recipe looks better now, thanks. But now the tests are failing because `ShortenFullyQualifiedTypeReferences` handles annotations unexpected. 

It seems to only shorten Annotations of the pattern ´packagename.@Annotation´ (see `ShortenFullyQualifiedTypeReferencesTest#typeFullyQualifiedAnnotatedField` and `ShortenFullyQualifiedTypeReferencesTest#annotatedFieldAccess`). 
In the `ShortenFullyQualifiedTypeReferences` fails if the annotation has the pattern `@packagename.TypeName`
According to the Java Language Spec 9.7.1 Annotations in FQN format have the pattern `@packagename.TypeName`. 
I think there might be a bug in `ShortenFullyQualifiedTypeReferences` or am I missing something?

Reply:
Quite possible that we have a bug there. I will try to investigate next week.